WIP: Complete overhaul of dockerfile

d2dyno opened the pull request:
There were a few issues here:
- Build was done inside the final image, rather than using multi-stage, to run build outside of the final image.
- fdk-aac was not the official source. (https://github.com/mstorsjo/fdk-aac is the official)
- FFMPEG was compiled non-statically, which meant a lot of packages had to be installed to the image. This version will grab all it needs from my ffmpeg docker, removing virtually all dependencies
- For simplicity, this is built with Ubuntu 18.04.

sandreas commented:
Nice work, thank you for the pull request. 

What I like:
- The separation of build and runtime image (I always wanted to do that)
- The official fdk aac source - i did not find it

Unfortunately there are also some things, that i would like to discuss. 

1.) Image does not build without a defined `${WORKDIR}`
I think the image should build without extra settings and use sane defaults.

2.) Ubuntu vs Alpine
The Alpine image is smaller - even with the build dependencies - so switching to ubuntu is not my preferred option - even if it is a minor difference

```
m4b-tool-d2dyno   215MB
m4b-tool          176MB
```

Well, I'll take a closer look at the improvements in the next days and will try to integrate them with alpine. Here is a sample how to do this: https://github.com/alfg/docker-ffmpeg/blob/master/Dockerfile

I'll leave the request open, till its done.

sandreas commented:
Well, today I tried to use the official fdkaac-repository but it did not compile successfully...

Here is a first try of an improved Dockerfile using alpine as base image with build image - but i did not check if everything works as expected.

```
###############################
# Build the FFmpeg-build image.
FROM alpine:3.8 as build
 
ARG FFMPEG_VERSION=4.1.2
 
ARG PREFIX=/opt/ffmpeg
ARG LD_LIBRARY_PATH=/opt/ffmpeg/lib
ARG MAKEFLAGS="-j4"
 
# FFmpeg build dependencies.
RUN apk add --update \
  autoconf \
  automake \
  build-base \
  coreutils \
  freetype-dev \
  gcc \
  libtool \
  lame-dev \
  libogg-dev \
  libvorbis-dev \
  libwebp-dev \
  libtheora-dev \
  opus-dev \
  pkgconf \
  pkgconfig \
  rtmpdump-dev \
  wget \
  yasm
 
# Get fdk-aac from testing.
RUN echo http://dl-cdn.alpinelinux.org/alpine/edge/testing >> /etc/apk/repositories && \
  apk add --update fdk-aac-dev
 
# Get ffmpeg source.
RUN cd /tmp/ && \
  wget http://ffmpeg.org/releases/ffmpeg-${FFMPEG_VERSION}.tar.gz && \
  tar zxf ffmpeg-${FFMPEG_VERSION}.tar.gz && rm ffmpeg-${FFMPEG_VERSION}.tar.gz
 
# Compile ffmpeg.
RUN cd /tmp/ffmpeg-${FFMPEG_VERSION} && \
  ./configure \
  --enable-version3 \
  --enable-gpl \
  --enable-nonfree \
  --enable-small \
  --enable-libmp3lame \
  --enable-libtheora \
  --enable-libvorbis \
  --enable-libopus \
  --enable-libfdk-aac \
  --enable-libwebp \
  --enable-librtmp \
  --enable-libfreetype \
  --enable-openssl \
  --disable-debug \
  --disable-doc \
  --disable-ffplay \
  --extra-cflags="-I${PREFIX}/include" \
  --extra-ldflags="-L${PREFIX}/lib" \
  --extra-libs="-lpthread -lm" \
  --prefix="${PREFIX}" && \
  make && make install
 
RUN echo "---- COMPILE SANDREAS MP4V2 ----" && \
    cd /tmp/ \
    && wget https://github.com/sandreas/mp4v2/archive/master.zip \
    && unzip master.zip \
    && rm master.zip \
    && cd mp4v2-master \
    && ./configure && make && make install
RUN echo "---- COMPILE FDKAAC ----" \
    && cd /tmp/ \
    && wget https://github.com/nu774/fdkaac/archive/1.0.0.tar.gz \
    && tar xzf 1.0.0.tar.gz \
    && rm 1.0.0.tar.gz \
    && cd fdkaac-1.0.0 \
    && autoreconf -i && ./configure && make && make install && rm -rf /tmp/*
 
# Cleanup.
#RUN rm -rf /var/cache/apk/* /tmp/*
 
FROM alpine:3.9.2
 
ENV WORKDIR /mnt
ARG M4B_TOOL_DOWNLOAD_LINK="https://github.com/sandreas/m4b-tool/releases/latest/download/m4b-tool.phar"
ARG FFMPEG_VERSION=4.1
ARG PREFIX=/ffmpeg_build
 
RUN echo "---- INSTALL RUNTIME PACKAGES ----" \
    && apk add --no-cache --update --upgrade bzip2 \
    ca-certificates \
    curl \
    file \
    freetype \
    lame \
    libogg \
    libvorbis \
    libtheora \
    libvorbis \
    openssl \
    opus \
    pcre \
    php7-cli \
    php7-dom \
    php7-intl \
    php7-json \
    php7-xml \
    php7-curl \
    php7-mbstring \
    php7-simplexml \
    php7-phar \
    && echo "date.timezone = UTC" >> /etc/php7/php.ini
COPY --from=build /opt/ffmpeg /usr/local/bin/fdkaac /usr/local/bin/mp4art /usr/local/bin/mp4chaps \
/usr/local/bin/mp4extract /usr/local/bin/mp4file /usr/local/bin/mp4info \
/usr/local/bin/mp4subtitle /usr/local/bin/mp4tags \
/usr/local/bin/mp4track /usr/local/bin/mp4trackdump \
/usr/local/bin/
COPY --from=build /usr/lib/libfdk-aac.so.2 /usr/lib/libfdk-aac.so.2
 
# workaround to copy a local m4b-tool.phar IF it exists
ADD ./Dockerfile ./dist/m4b-tool.phar* /tmp/
RUN echo "---- INSTALL M4B-TOOL ----" \
    && if [ ! -f /tmp/m4b-tool.phar ]; then wget "${M4B_TOOL_DOWNLOAD_LINK}" -O /tmp/m4b-tool.phar ; fi \
    && mv /tmp/m4b-tool.phar /usr/local/bin/m4b-tool \
    && chmod +x /usr/local/bin/m4b-tool
 
WORKDIR ${WORKDIR}
CMD ["list"]
ENTRYPOINT ["m4b-tool"]
```

d2dyno commented:
I wasn't sure if you were set on Alpine or not. Knowing you would like it Alpine-based, I will re-base the file.

sandreas commented:
> I wasn't sure if you were set on Alpine or not.

Well, not set to Alpine but i would like to provide the smallest possible and still functional image size - i think alpine is the way to go...

>  Are you also set on building FFMPEG within this dockerfile?

No. I integrated a custom ffmpeg build mainly because of `libfdk_aac` and on my way i optimized the ffmpeg build removing stuff like `h264` and `h265` that will never be needed for `m4b-tool`. But if the `ffmpeg` is not MUCH bigger in size, i would prefer using an existing config instead of using my own.

Well, every pull request is welcome and i appreciate your work. If the image size is < 200MB and using alpine, i would accept it right away... Perhaps it would be also nice to publish the image at dockerhub, so that it is available via docker pull...

d2dyno commented:
Please note the WIP added, as I have not run any tests on this yet.

I have re-based my work on alpine. Image built on my end is now `56.43 MiB`

sandreas commented:
Thank you very much. I'll take a closer look and run my tests on your image in the next days and lift it into my production environment to do further testing if it works.

I really appreciate your work :-)

d2dyno commented:
Tested a bit and updated dockerfile accordingly. Also added a build for you to test via dockerhub https://hub.docker.com/r/d2dyno/m4b-tool (might still be building)

sandreas commented:
Thank you...  I'll take a look in the next days

sandreas commented:
Ok, i took a first look and there are many improvements over the original image - but unfortunately some things are missing. 

I'm working on a fix and try to provide a Dockerfile for discussion...

sandreas commented:
Some further notes:
- `mp4art` (and other mp4v2 tools) are not working because libs are missing (at least `libmp4v2.so.2`)
- The fdkaac *official repository* does not contain the binary encoder - so using `nu774/fdkaac` was correct and it has to be included (see below !!!!! on the apt-cache dump on recent ubuntu for details)
- A part of the Dockerfile, where a local build of `m4b-tool.phar` is preferred only, when its available, is missing

BUT: I would like to thank you, for your work so far. I'm on it, if it is working, I'll let you know...

```
❯ apt-cache show fdkaac
Package: fdkaac
Architecture: amd64
Version: 0.6.3-1
Priority: optional
Section: multiverse/sound
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Marius Gavrilescu <marius@ieval.ro>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 104
Depends: libc6 (>= 2.14), libfdk-aac1 (>= 0.1.4)
Filename: pool/multiverse/f/fdkaac/fdkaac_0.6.3-1_amd64.deb
Size: 39154
MD5sum: c821013f26def4ac39175e9321c9e3ce
SHA1: 197bdd79ea24d11620a29a83251549fd0c4e2482
SHA256: 8d222c8047b3c13d7f37b8610787156f5f7cd58d490cdae6e720131d0d8f35b2
!!!!! Homepage: https://github.com/nu774/fdkaac
Description-en: command line encoder frontend for libfdk-aac
 fdkaac is a command-line AAC encoder based on libfdk-aac.
 .
 It reads linear PCM audio in either WAV, raw PCM, or CAF format and
 encodes it into M4A / AAC files.
Description-md5: fcbe7317aa8b9e391239fe1281b1c42a
```

sandreas commented:
Here's a version that should work and i think that it can hardly be improved or stripped down any more... (except compiling `ffmpeg` with only these dependencies, `m4b-tool` needs) 
To give a better understanding, what happens and why the dependencies are needed, i would like to add the following explanations:

- The single `RUN` statement instead of multiple ones keeps the `build` image small, because it has only one layer
- The `--no-cache` keeps the image small, because apk-files are not stored permanentely
- The `--virtual=build-dependencies` keeps the image small, because it is possible to remove all build dependencies by removing one package after the build
- `m4b-tool` requires `mp4v2`, `ffmpeg` with `libfdk-aac`, `fdkaac` encoder (binary, not only lib)
- `mp4v2` is a patched version by myself to support properties for sorting (e.g. `sort-title`) which is not part of the official `mp4v2` command line tools
- `ffmpeg` with `libfdk-aac` is available as *static* version (that means, it is a big single binary without any dependencies) - this is done via `COPY --from=mwader/static-ffmpeg:4.1.3-1 /ffmpeg /usr/local/bin/` (i used it, because the non static version had problems with crypt
- The shell `if ... curl ...` scripting is for using a self compiled version instead of the stable version, if a file `./dist/m4b-tool.phar` exists that has been compiled before... it is needed

```
##############################
#
#   m4b-tool build image
#
##############################
FROM alpine:3.9.2 as build
RUN echo "---- INSTALL BUILD DEPENDENCIES ----" \
    && echo http://dl-cdn.alpinelinux.org/alpine/edge/testing >> /etc/apk/repositories \
    && apk add --no-cache --update --upgrade --virtual=build-dependencies \
    autoconf \
    automake \
    boost-dev \
    build-base \
    gcc \
    git \
    tar \
    fdk-aac-dev \
    wget && \
echo "---- COMPILE SANDREAS MP4V2 (for sort-title, sort-album and sort-author) ----" \
  && cd /tmp/ \
  && wget https://github.com/sandreas/mp4v2/archive/master.zip \
  && unzip master.zip \
  && cd mp4v2-master \
  && ./configure && \
  make -j4 && \
  make install && make distclean && \
echo "---- COMPILE FDKAAC ENCODER (executable binary for usage of --audio-profile) ----" \
    && cd /tmp/ \
    && wget https://github.com/nu774/fdkaac/archive/1.0.0.tar.gz \
    && tar xzf 1.0.0.tar.gz \
    && cd fdkaac-1.0.0 \
    && autoreconf -i && ./configure && make -j4 && make install && rm -rf /tmp/* && \
echo "---- REMOVE BUILD DEPENDENCIES (to keep image small) ----" \
   && apk del --purge build-dependencies && rm -rf /tmp/*

##############################
#
#   m4b-tool runtime image
#
##############################
FROM alpine:3.9.2

ENV WORKDIR /mnt
ARG M4B_TOOL_DOWNLOAD_LINK="https://github.com/sandreas/m4b-tool/releases/latest/download/m4b-tool.phar"

RUN echo "---- INSTALL RUNTIME PACKAGES ----" && \
  apk add --no-cache --update --upgrade \
  # mp4v2: required libraries
  libstdc++ \
  # m4b-tool: php cli, required extensions and php settings
  php7-cli \
  php7-dom \
  php7-json \
  php7-xml \
  php7-mbstring \
  php7-phar \
  php7-tokenizer \
  php7-xmlwriter \
  && echo "date.timezone = UTC" >> /etc/php7/php.ini \
  # fdkaac: required libaries
  && echo http://dl-cdn.alpinelinux.org/alpine/edge/testing >> /etc/apk/repositories \
      && apk add --no-cache --update --upgrade fdk-aac-dev


# copy ffmpeg static with libfdk from mwader docker image
COPY --from=mwader/static-ffmpeg:4.1.3-1 /ffmpeg /usr/local/bin/

# copy compiled mp4v2 binaries, libs and fdkaac encoder to runtime image
COPY --from=build /usr/local/bin/mp4* /usr/local/bin/fdkaac /usr/local/bin/
COPY --from=build /usr/local/lib/libmp4v2* /usr/local/lib/


# workaround to copy a local m4b-tool.phar IF it exists
ADD ./Dockerfile ./dist/m4b-tool.phar* /tmp/
RUN echo "---- INSTALL M4B-TOOL ----" \
    && if [ ! -f /tmp/m4b-tool.phar ]; then curl -o /tmp/m4b-tool.phar "${M4B_TOOL_DOWNLOAD_LINK}"; fi \
    && mv /tmp/m4b-tool.phar /usr/local/bin/m4b-tool \
    && chmod +x /usr/local/bin/m4b-tool


WORKDIR ${WORKDIR}
CMD ["list"]
ENTRYPOINT ["m4b-tool"]
```

sandreas commented:
Well, I really appreciate your work and like it, that you tried to improve the `Dockerfile` with your ideas - but this pull request is still open and i did not receive a response yet, so i would like to ask, if this is still a work in progress or if you would like to add something?

I was not able to reach a real improvement with the customizations, you did, what do you think?

sandreas commented:
Since there is no progress in this pull request, and due to conflicts the merge is not possible, i will close this pull request for now. If you would like to add something or suggest further improvements, i would appreciate every progress.
